File: backend/libbackend/file.ml
open Core
open Libcommon
open Lwt.Infix
open Libexecution

(* ------------------- *)
(* file stuff with checks *)
(* ------------------- *)
let check_filename ~(root : Config.root) ~mode f =
  let dir = Config.dir root in
  let f = dir ^ f in
  let debug name value =
    if false then Log.debuG name ~data:(string_of_bool value) ;
    value
  in
  if root <> No_check
     && ( String.is_substring ~substring:".." f |> debug "dots"
        || String.contains f '~' |> debug "tilde"
        || String.is_suffix ~suffix:"." f |> debug "tilde"
        || (mode <> `Dir && String.is_suffix ~suffix:"/" f)
           |> debug "ends slash"
        || (not (String.is_suffix ~suffix:"/" dir)) |> debug "dir no slash"
        || String.is_substring ~substring:"etc/passwd" f |> debug "etc"
        (* being used wrong *)
        || String.is_substring ~substring:"//" f |> debug "double slash"
        (* check for irregular file *)
        || (mode = `Read && not (Sys.is_file f = `Yes)) )
        |> debug "irreg"
  then (
    Log.erroR "SECURITY_VIOLATION" ~data:f ;
    Exception.internal "FILE SECURITY VIOLATION" )
  else f


let file_exists ~root f : bool =
  let f = check_filename ~root ~mode:`Check f in
  Sys.file_exists f = `Yes


let mkdir ~root dir : unit =
  let dir = check_filename ~root ~mode:`Dir dir in
  Unix.mkdir_p dir


let lsdir ~root dir : string list =
  let dir = check_filename ~root ~mode:`Dir dir in
  Sys.ls_dir dir


let rm ~root file : unit =
  let file = check_filename ~root ~mode:`Write file in
  Core_extended.Shell.rm ~f:() file


let readfile ~root f : string =
  let f = check_filename ~root ~mode:`Read f in
  let ic = Caml.open_in f in
  try
    let n = Caml.in_channel_length ic in
    let s = Bytes.create n in
    Caml.really_input ic s 0 n ;
    Caml.close_in ic ;
    Caml.Bytes.to_string s
  with e ->
    Caml.close_in_noerr ic ;
    raise e


let readfile_lwt ~root f : string Lwt.t =
  let f = check_filename ~root ~mode:`Read f in
  Lwt_io.with_file ~mode:Lwt_io.input f Lwt_io.read


let writefile ~root (f : string) (str : string) : unit =
  let f = check_filename ~root ~mode:`Write f in
  let flags = [Unix.O_WRONLY; Unix.O_CREAT; Unix.O_TRUNC] in
  Unix.with_file ~perm:0o600 ~mode:flags f ~f:(fun desc ->
      ignore (Unix.write desc ~buf:(Bytes.of_string str)) )


(* ------------------- *)
(* json *)
(* ------------------- *)

let readjsonfile
    ~root
    ?(stringconv : string -> string = ident)
    ~(conv : Yojson.Safe.t -> ('a, string) result)
    (filename : string) : 'a =
  filename
  |> readfile ~root
  |> stringconv
  |> Yojson.Safe.from_string
  |> conv
  |> Result.ok_or_failwith


let maybereadjsonfile
    ~root
    ?(stringconv : string -> string = ident)
    ~(conv : Yojson.Safe.t -> ('a, string) result)
    (filename : string) : 'a option =
  if file_exists ~root filename
  then Some (readjsonfile ~root ~stringconv ~conv filename)
  else None


(* ------------------- *)
(* spawning *)
(* ------------------- *)

let init () =
  (* Spawn creates lots of child processes. When they finish, the OS
   * asks the dark executable what to do. This tells it to ignore them
   * in such a way that the OS will clean them up. (I thought this was
   * the default, but this appears to fix the problem). *)
  Signal.ignore Signal.chld

File: backend/src/LibBackend/File.fs
/// Helpers to work with files on disk
module LibBackend.File

// This makes extra careful that we're only accessing files where we expect to
// find files, and that we're not checking outside these directories

// Note: none of these are async because System.IO is not async

open Prelude
open Tablecloth

type Mode =
  | Check
  | Dir
  | Read
  | Write

let checkFilename (root : Config.Root) (mode : Mode) (f : string) =
  let dir = Config.dir root
  let f : string = $"{dir}{f}"

  let debug (name : string) (value : bool) =
    if value then print $"checkFilename failed: {name}: {value}"
    value

  if
    (root <> Config.NoCheck)
    && (f.Contains ".." |> debug "dots"
        || f.Contains "~" |> debug "tilde"
        || f.EndsWith "." |> debug "ends dot"
        || (mode <> Dir && f.EndsWith "/") |> debug "ends slash"
        || (not (dir.EndsWith "/")) |> debug "dir no slash"
        || f.EndsWith "etc/passwd" |> debug "etc"
        // being used wrong
        || f.EndsWith "//" |> debug "double slash"
        // check for irregular file
        || (mode = Read
            && (System.IO.File.GetAttributes f <> System.IO.FileAttributes.Normal)
            && (System.IO.File.GetAttributes f <> System.IO.FileAttributes.ReadOnly))
           |> debug "irreg")
  then
    Exception.raiseInternal "FILE SECURITY VIOLATION" [ "file", f ]
  else
    f


let fileExists root f : bool =
  let f = checkFilename root Check f
  System.IO.File.Exists f

let lsdir (root : Config.Root) (dir : string) : string list =
  let absoluteDir = checkFilename root Dir dir

  absoluteDir
  |> System.IO.Directory.EnumerateFileSystemEntries
  |> Seq.toList
  |> List.map (String.dropLeft absoluteDir.Length)

let lsPattern (root : Config.Root) (pattern : string) : string list =
  let absoluteDir = checkFilename root Dir ""

  System.IO.Directory.EnumerateFileSystemEntries(absoluteDir, pattern)
  |> Seq.toList
  |> List.map (String.dropLeft absoluteDir.Length)


let readfile (root : Config.Root) (f : string) : string =
  f |> checkFilename root Read |> System.IO.File.ReadAllText

let readfileBytes (root : Config.Root) (f : string) : byte[] =
  f |> checkFilename root Read |> System.IO.File.ReadAllBytes

let tryReadFile (root : Config.Root) (f : string) : string option =
  if fileExists root f then
    f |> checkFilename root Read |> System.IO.File.ReadAllText |> Some
  else
    None

let rec writefileBytes (root : Config.Root) (f : string) (contents : byte[]) : unit =
  let f = checkFilename root Write f

  // First write to a temp file, then copy atomically. Do this as we've lost our data
  // a few times.
  let tempFilename = System.IO.Path.GetTempFileName()
  System.IO.File.WriteAllBytes(tempFilename, contents)

  // We might not be the only one trying to copy here, and .NET won't let us
  // overwrite it if something else is. So try again.
  let mutable success = false
  let mutable count = 0
  while success = false && count < 10 do
    try
      System.IO.File.Move(tempFilename, f, true)
      success <- true
    with e ->
      count <- count + 1
      if count > 10 then e.Reraise() else ()
      ()

let rec writefile (root : Config.Root) (f : string) (contents : string) : unit =
  writefileBytes root f (UTF8.toBytes contents)
